torch/csrc/distributed/Module.cpp

```diff
@@ -338,6 +367,271 @@ PyObject* THDPModule_recv(PyObject *_unused, PyObject *args)
+
+  descriptors.reserve(length);
+
+  for (std::size_t i = 0; i < length; ++i) {
+    if (!THPModule_isTensor(PySequence_Fast_GET_ITEM(sequence, i))) {
+      goto invalid_arguments;
+    }
+
+    descriptors.push_back(
+      THDPModule_makeDescriptor(PySequence_Fast_GET_ITEM(sequence, i))
+    );
+  }
+
+  group = _getGroup(PyTuple_GET_ITEM(args, 2));
+  op = _getReduceOp(PyTuple_GET_ITEM(args, 1));
+
+  {
+    AutoNoGIL guard;
+    THDAllReduceMultiGPU(descriptors.data(), length, op, group);
+  }
+  Py_RETURN_NONE;
+
+invalid_arguments:
+  THPUtils_invalidArguments(args, NULL, "all_reduce_multigpu", 1,
+                            "(list[tensor] in_out, reduce_op op, group gr)");
+  Py_RETURN_NONE;
+  END_HANDLE_TH_ERRORS
+}
+
+
+PyObject* THDPModule_reduceMultiGPU(PyObject *_unused, PyObject *args)
+{
+  HANDLE_TH_ERRORS
+  PyObject* sequence;
+  Py_ssize_t tmp_length;
+  std::size_t length;
+  std::vector<at::Tensor> descriptors;
+  THDGroup group;
+  THDReduceOp op;
+  int dst_rank;
+
+  if (PyTuple_GET_SIZE(args) != 4) {
+    goto invalid_arguments;
+  }
+
+  if (!PySequence_Check(PyTuple_GET_ITEM(args, 0)) ||
+      !THPUtils_checkLong(PyTuple_GET_ITEM(args, 1))) {
+    goto invalid_arguments;
+  }
+
+  sequence = PySequence_Fast(PyTuple_GET_ITEM(args, 0), nullptr);
+  if (!sequence) {
+    goto invalid_arguments;
+  }
+
+  tmp_length = PySequence_Fast_GET_SIZE(sequence);
+  THPUtils_assert(tmp_length >= 0, "couldn't obtain the length of %s",
+                  THPUtils_typename(sequence));
+  length = static_cast<std::size_t>(tmp_length);
+
+  descriptors.reserve(length);
+
+  for (std::size_t i = 0; i < length; ++i) {
+    if (!THPModule_isTensor(PySequence_Fast_GET_ITEM(sequence, i))) {
+      goto invalid_arguments;
+    }
+
+    descriptors.push_back(
+      THDPModule_makeDescriptor(PySequence_Fast_GET_ITEM(sequence, i))
+    );
+  }
+
+  group = _getGroup(PyTuple_GET_ITEM(args, 3));
+  op = _getReduceOp(PyTuple_GET_ITEM(args, 2));
+  dst_rank = THPUtils_unpackLong(PyTuple_GET_ITEM(args, 1));
+
+  {
+    AutoNoGIL guard;
+    THDReduceMultiGPU(descriptors.data(), length, op, dst_rank, group);
+  }
+  Py_RETURN_NONE;
+
+invalid_arguments:
+  THPUtils_invalidArguments(args, NULL, "reduce_multigpu", 1,
+                            "(list[tensor] in_out, int dst_rank, "
+                            "reduce_op op, group gr)");
+  Py_RETURN_NONE;
+  END_HANDLE_TH_ERRORS
+}
+
+
+PyObject* THDPModule_broadcastMultiGPU(PyObject *_unused, PyObject *args)
+{
+  HANDLE_TH_ERRORS
+  PyObject* sequence;
+  Py_ssize_t tmp_length;
+  std::size_t length;
+  std::vector<at::Tensor> descriptors;
+  THDGroup group;
+  int src_rank;
+
+  if (PyTuple_GET_SIZE(args) != 3) {
+    goto invalid_arguments;
+  }
+
+  if (!PySequence_Check(PyTuple_GET_ITEM(args, 0)) ||
+      !THPUtils_checkLong(PyTuple_GET_ITEM(args, 1))) {
+    goto invalid_arguments;
+  }
+
+  sequence = PySequence_Fast(PyTuple_GET_ITEM(args, 0), nullptr);
```

Comment:
I'm not 100% sure if error message can be NULL. The docs don't say that it can, so I'd assume it shouldn't. Just put a simple error message saying that this function expected a sequence.